cmd/dockerd: move code for managed containerd to unix-only files

thaJeztah opened the pull request:
Starting containerd as managed child-process is only used on Linux,
so we can move this code to linux-specific files.

This also refactors the DaemonCLI.initContainerd();

- return early if we're expecting a system-containerd
- handle WaitTimeout() errors by logging them
- move creation of options inline, now that we no longer
  need platform-specific abstractions.

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
cmd/dockerd: move code for managed containerd to unix-only files

Starting containerd as managed child-process is only used on Linux,
so we can move this code to linux-specific files.

This also refactors the DaemonCLI.initContainerd();

- return early if we're expecting a system-containerd
- handle WaitTimeout() errors by logging them
- move creation of options inline, now that we no longer
  need platform-specific abstractions.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

tianon commented:
This gets more complicated if/as we look to transition Windows to use containerd-by-default, right?  Would we expect users to manually set up a separate Windows service for containerd? IMO that's the _best_ option, but I don't know how realistic that is - probably fine, especially since we don't have actual "packages" on Windows, just an item to put in the release notes for the release that makes that switch?

thaJeztah commented:
Yeah, so this PR does not change things from a functionality perspective, but just removes code from platforms where it's not used.

As to "complicated to install containerd separately", that's a fair point. I know that when the containerd integration was merged for Windows, there was some discussion wether or not "running containerd as child-process" should be deprecated altogether (which was I think the motivation of not implementing it for Windows).

I seem to recall Mirantis already uses containerd on Windows for running kubernetes (I may be wrong), but perhaps @corhere knows how their installation procedure is to install and run containerd and docker on Windows

(For Docker Desktop, I assume we can make the installation and setup part of the general installation process for docker desktop on Windows).

corhere commented:
I do not know whether Mirantis products use containerd on Windows, though our resident build & packaging guru @ameyag might.

ameyag commented:
UCP/MKE uses containerd on Windows. A general install script performs installation and setup for containerd as a follow-up to a successful setup of docker service. We can modify it to install containerd before docker. Is this going to be for 22.06?

tianon commented:
I'll let @thaJeztah contradict me if I'm speaking out of turn, but I think this is _not_ for 22.06 (which is probably part of why he's still got it as a draft). :+1:

thaJeztah commented:
> I'll let @thaJeztah contradict me if I'm speaking out of turn, but I think this is _not_ for 22.06 (which is probably part of why he's still got it as a draft). 👍

Overall, the change itself should be "fine" for 22.06, as it should not result in a change of behavior (only removing code from platforms that don't use it). Mostly opened it as draft as it was not _super_ important, and I wanted to have a run of CI 😂

corhere commented on cmd/dockerd/daemon.go:
```diff
@@ -172,14 +171,12 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
-	if waitForContainerDShutdown != nil {
-		defer waitForContainerDShutdown(10 * time.Second)
-	}
+	waitForContainerdShutdown, err := cli.initContainerd(ctx)
 	if err != nil {
 		cancel()
 		return err
 	}
+	defer waitForContainerdShutdown()
 	defer cancel()
```
Pet peeve:
```suggestion
	defer cancel()
	waitForContainerdShutdown, err := cli.initContainerd(ctx)
	if err != nil {
		return err
	}
	defer waitForContainerdShutdown()
```

thaJeztah replied:
Ah, yes, I was looking at that one, and wasn't sure if the order in which they're called was important 🤔

thaJeztah replied:
done

corhere commented on cmd/dockerd/daemon_unix.go:
```diff
@@ -130,33 +122,50 @@ func newCgroupParent(config *config.Config) string {
+		return nil, errors.Wrap(err, "failed to start containerd")
+	}
+	cli.Config.ContainerdAddr = r.Address()
+	logrus.Info("started daemon managed containerd using " + cli.Config.ContainerdAddr)
```
Nit: structured logging?
```suggestion
	logrus.WithField("address", cli.config.ContainerdAddr).Info("started daemon managed containerd")
```

thaJeztah replied:
done

cpuguy83 commented:
I would honestly like to move away from managing a separate process. I understand there are some trade offs with that.

cpuguy83 commented:
More specifically I'm talking about spinning up containerd services in process.

thaJeztah commented:
> I would honestly like to move away from managing a separate process. I understand there are some trade offs with that.

Yeah, we should look at that. That said; it's super convenient for testing to be able to do a quick `./dockerd --debug` (definitely use that all the time when developing/testing). I guess we could have  a script for that to make that still possible. Overall, I expect not many use this outside of running docker-in-docker, or for "playground", so not sure how much of a concern it is (I think the amount of code related is limited)

thaJeztah committed:
revert defer changes

Seeing lots of tests fail with;

    [d941d25b030a6] error while stopping the daemon: exit status 2

So checking if it's related

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
